### ISSUE TITLE
Add support for PSR7 StreamInterface

### ISSUE THREAD

**Issue by phaldan** [pcrov/JsonReader#10]
Is it possible to extend the API with a new method to use a PSR7 StreamInterface.
This would make it easier to parse the JSON from the request response with HTTP Clients like Guzzle.
My currenty workarround contains a [StreamWrapper](http://php.net/manual/en/class.streamwrapper.php) to map from a StreamInterface to a resource and it took me a while to find the following solution:

```php
use GuzzleHttp\Psr7\StreamWrapper;
use pcrov\JsonReader\JsonReader;

$stream = $response->getBody();
$resource = StreamWrapper::getResource($stream);
$reader = new JsonReader();
$reader->stream($resource);
```

This could be realised with a new InputStream implementation for the PSR7 StreamInterface.

**Comment by pcrov** [comment]
I don't see why not. Any preference as to the name of the method? (Always the hard part. Hell, I'm sorta tempted to just allow them directly in JsonReader::stream().)

As an aside, it's already possible to use a custom InputStream (which are dead simple things, a `read` method that spits out strings then null on eof) via JsonReader::init(), e.g.:
```php
$reader->init(new JsonParser(new Lexer($yerInputStream)));
```
I should probably document that better.

**Comment by pcrov** [comment]
I just pushed a feature branch for this, psr7stream. It adds a new InputStream and allows PSR7 streams in JsonReader::stream(). Care to give it a quick smoke test? If everything looks good I'll add tests and merge it for release.

**Comment by phaldan** [comment]
I realized after I have seen your adapter pattern with the `InputStream`, that I could have defined my own `InputStream` for the PSR7 StreamInterface. This small, but important detail, should definitely be document with an example. Other libraries for a JSON stream reader doesn't abstract their input reader.

Your solution works, but wouldn't it be a cleaner solution to avoid different data types for the same parameter and instead use type hinting with a new method like `psr7Stream(...)` or `httpStream(...)`:
```php
public function httpStream(StreamInterface $stream): void
{
  $input = new Psr7Stream($stream);
  $reader->init(new JsonParser(new Lexer($input)));
}
```

**Comment by pcrov** [comment]
`psr7Stream` works for me, branch updated. I'll try to get to adding tests later today. I've also opened #11 so I don't forget to document using custom InputStreams.

**Comment by pcrov** [comment]
Just tagged a release for this, [0.8.0](https://github.com/pcrov/JsonReader/releases/tag/0.8.0). This might turn into 1.0 soon, so if you've got any other suggestions or run into any issues please let me know.

**Comment by phaldan** [comment]
I already used your library to read a lot of small json files (a few 100k in total with a size arround a few KB) and had zero issues. Some bigger files with a few MB in size were also among them. I could not detect any memory leaks and or wrong readings. 

A thing I thought about would be nice to have are Static Factory Methods beside already existing instance methods for the input. Like the following:
```php
class JsonReader
{
  //...
  public static method fromPsr7Stream(StreamInterface $stream, int $options = 0): JsonReader
  {
     $reader = new JsonReader($options);
     return $reader->psr7Stream($stream); // Also Fluent API would ne nice
  }
```

Another thing would be to have a Fluent API to allow method chaining and is there a reason why you didn't specify a return type in your public methods?

**Comment by pcrov** [comment]
I'm likely swapping to named constructors much like that in v2. Originally this lib targeted people migrating from dealing with large XML documents to large JSON ones, so I mimicked XMLReader very closely. I'd rather swap than have both methods of initialization public, just to keep down clutter, and it's too late in the cycle toward v1 for that kind of change to JsonReader's API. The less people have to do to get on a stable version the better.

I'm not particularly keen on fluent APIs, largely for the [same reasons as Ocramius](https://ocramius.github.io/blog/fluent-interfaces-are-evil/). I'd consider them for something akin to xpath or xquery, but that'd be its own thing. (What I really want is a DSL that I don't hate targeting JSON, but this has yet to surface.)

This lib currently (and for v1) supports PHP 7.0 so many return types are not available (e.g. void, nullables.) They're going to be added for v2 which'll bump the minimum PHP version to whatever is latest at the time. If I've missed any others that's a bug.

Glad to hear you haven't run into any issues. I am curious though why you're not using json_decode for small (or even mid-sized) documents, as it should be quite a bit faster. The only edge cases that come to mind for using JsonReader instead are things like duplicate object member names or to handle floats with BCMath. Are you in a super memory-constrained environment?

Not trying to put you off using this or anything, just wondering if there's another use case in the wild I should keep in mind going forward.

**Comment by phaldan** [comment]
I totally unstead that you want to keep it simple. It was only me trying to save some lines and this was also the main idea behind the fluent api.

I totally forgot that `void` was not shipped with PHP 7.0. It's such an essential return type.

I start using your library, because I had to read an index file with arround 5MB, that is formated as json. Based on the index I read additional information for each entry. I only need one entry at a time, so my idea was to use an iterator. A stream reader in this case is a memory efficent solution for an iterator.

It was not required to have highly optimized memory solution, but now my CLI script has more or less zero impact on the memory usage. Also I didn't liked the idea to pass arround a json-string or some generic `stdClass`/`array` and it's to much overhead to create models for the additional information. Indeed your solution is a little bit slow then `json_decode`, but in my case is only 10% slower to read the additional information.
